### DISCUSSION TITLE
How to pass default value to unmapped constructor parameter

### DISCUSSION THREAD

**Discussion by KUTlime** [riok/mapperly#335] [Q&A]
Imagine this repository entity:

```csharp
public record SomeEntity(long Id, Guid UniqueId, string SomeProperty);
```
and API model:

```csharp
public class SomeAPIModel
{
   public Guid UniqueId {get; set;} // A public unique ID 

   public string SomeProperty {get; set;}
}
```

Using Mapperly, mapping SomeEntity -> SomeAPIModel is fine, because Mapperly will map 3 properties to 2 properties, not vice versa. From SomeAPIModel -> SomeEntity, we missing value for Id. The use of `default` would be OK, because the value will overridden anyway, but generation of mapping would fail on this.

Is there any way how to overcome this? How to pass a value for certain properties?

**Comment by latonz** [comment]
What compiler error is thrown?

You should be able to ignore the property by using the `MapperIgnoreTarget` attribute. See [docs](https://mapperly.riok.app/docs/configuration/mapper#properties).

**Reply by KUTlime** [reply]
Dear @latonz thanks for your reply, I've updated my question. I check the entity, it is a record with generated constructor.

The errors are translated into my language (_Czech_), so I'm pasting MVE:

```csharp
using Riok.Mapperly.Abstractions;

public record SomeEntity(long Id, Guid UniqueId, string SomeProperty);

public class SomeAPIModel
{
    public Guid UniqueId { get; set; } // A public unique ID

    public string SomeProperty { get; set; }
}

[Mapper]
public static partial class Mapper
{
    public static partial SomeEntity ToEntity(SomeAPIModel model);
}
```

This will fail on build.

**Reply by latonz** [reply]
Got it, the error you are receiving is probably `RMG013 No accessible constructor with mappable arguments found`. This is the case since Mapperly cannot map to the ctor of `SomeEntity` as it does not know how to map to the `id` parameter. Either add another ctor without the `id` parameter or move the `id` parameter to the last position and add a default value.

**Reply by KUTlime** [reply]
Dear @latonz if I image a parallel universe in which I can push trough a code review a repository entity with Id as optional parameter placed on the end with reasoning "It is because of 3rd party source code generated mapping.", it is a really good or ideal solution, right?

I think it shouldn't be a problem to pass a default value to the mapping from Mapperly (_which is basically what do you suggest, just more versatile and clear architecture_), not bend the entity that is being mapped.

**Reply by latonz** [reply]
I don't think the solution is to let Mapperly inject default arguments into unknown constructor parameters. Also I can't think of a deserialiser (eg. `System.Text.Json`) doing this. IMO this would not be reasonable.
You always have the option to implement a specific mapping yourself or make use of an object factory.

**Reply by KUTlime** [reply]
Why unknown? 

What is wrong about having an option like this? (_name proposals_)
```csharp
[SetValueForMember(nameof(SomeEntity.Id), 0)]
[PassValueToMember(nameof(SomeEntity.Id), 0)]
[SetMemberValue(nameof(SomeEntity.Id), 0)]
[InitiliazeMemberWithValue(nameof(SomeEntity.Id), 0)]
```

It would enable Mapperly to cover more scenarios That is what we want, right? To make it usable as much as possible.

**Reply by cremor** [reply]
A simpler fist step could be to have something like `[MapperSetDefaultValue(nameof(SomeEntity.Id))]`. The source generator could just always assign `default` to the property given in that attribute.

Or a possible solution that doesn't even need a new attribute: What if the existing `MapperIgnoreTargetAttribute` would just cause the source generator to assign `default` if it is a constructor parameter or required property?

**Reply by cremor** [reply]
I've created a feature request: #631

**Comment by cremor** [comment]
I would need something similar. Instead of missing constructor parameters, it's missing required members in my case (error message "RMG023	Required member A on mapping target type X was not found on the mapping source type Y").

I like the new attribute name proposals mentioned here: https://github.com/riok/mapperly/discussions/335#discussioncomment-5653264

**Reply by TimothyMakkison** [reply]
Have you tried using an [object factory](https://mapperly.riok.app/docs/configuration/object-factories)?

**Reply by cremor** [reply]
Yes, using an object factory is a workaround. But it's not really nice looking code because you need to initialize all properties in the object factory, even those that could be mapped from the source object. So this might result in many `Xyz = default` or (even worse) `Xyz = default!` assignments in the object factory.

**Comment by latonz** [comment]
Closing as there exists an issue now: https://github.com/riok/mapperly/issues/631
